### ISSUE TITLE
[java] False positive in ConsecutiveLiteralAppends with lambda expression

### ISSUE THREAD

**Issue by acanda** [pmd/pmd#561]
**Rule Set:**
PMD 5.8.1, rulesets/java/strings.xml/ConsecutiveLiteralAppends

**Description:**
PMD ignores the non-literal append in the lambda expression and reports `Rule:ConsecutiveLiteralAppends Priority:3 StringBuffer (or StringBuilder).append is called 2 consecutive times with literal Strings. Use a single append with a single combined String..` at the third line in the example below.

Plus there is an unnecessary second full stop at the end of the error message.

**Code Sample demonstrating the issue:**
```
final DoubleStream numbers = DoubleStream.generate(Math::random).limit(10);
final StringBuilder builder = new StringBuilder();
builder.append('[');
numbers.forEach(number -> builder.append(number));
builder.append(']');
```

**Running PMD through:**
maven-pmd-plugin 3.8 with pmd-java 5.8.1

**Comment by jsotuyod** [comment]
Definitely a false positive, and a quite interesting one. Since the lambdas are now stopper to keep looking downwards, we fail to see the append.

I'm not sure we can accurately detect this one... we would need to know for certain when the lambda is being executed... on this case that means knowing the semantics of `Stream::forEach`....

This can't even be fixed with cross-class CFG analysis, since on this case the class is part of the classpath (the JRE) and the source not available...

Short of a whitelist (with all the shortcomings of these), I can't think of a way around this for now... Ideas are most welcome!

**Comment by acanda** [comment]
You're absolutely right. If we replace
`numbers.forEach(number -> builder.append(number));` with `numbers.peek(number -> builder.append(number));` then it would be a valid violation since the append would never be executed.

What do you usually do in a case where you can't decide if it is a violation or not? Report the potential violation and suppress it with @SuppressWarnings? Or not report it at all?

**Comment by jsotuyod** [comment]
@acanda we don't have a formal policy. That would be a deeper discussion (do we prefer false positives or false negatives?). It would be interesting to analyze it and try to be more consistent.

**Comment by acanda** [comment]
If it were up to me I'd prefer false positives over false negatives. With false positives you are made aware that there is a potential problem and can either fix it if it's a true positive or suppress it if it's a false positive. You could even reduce the number of false positives with heuristics like the whitelist you mentioned.

With false negatives however you are not made aware and might miss a true violation.

In this specific case mixing StringBuilder and Stream is a code smell as the same can be done with the Stream API only:
```
        DoubleStream.generate(Math::random)
                    .limit(10)
                    .mapToObj(number -> Double.toString(number))
                    .collect(Collectors.joining("", "[", "]"));
```
There might even be a new rule in there: UseCollectJoiningInsteadOfStringBuilder ;-)

As for this issue I'd be happy if you close this as a "won't fix" since I think it's better to have false positives you can suppress than false negatives you are not aware of.

**Comment by jsotuyod** [comment]
Marking as won't fix. I'd love to see a proposal (or even better, a PR!) for `UseCollectJoiningInsteadOfStringBuilder`

Thanks for your input!
